language/sygus.tex

```diff
@@ -1068,29 +1083,62 @@ \subsection{Asserting Synthesis Constraints and Assumptions}
 where $(x_1, \ldots, x_n)$ and $(y_1, \ldots, y_n)$
 are tuples of variables of sort $(\sigma_1, \ldots, \sigma_n)$ and
 $\varphi_{pre}$, $\varphi_{trans}$ and $\varphi_{post}$ are formulas.
-Given that this command is well-formed, given the above definitions,
+
+When this command is well-formed, given the above definitions,
 this command is syntax sugar for:
-\[
-\begin{array}{l}
-   \paren{\vardeclkwd\mbox{ }v_1\mbox{ }\sigma_1}\\
-   \paren{\vardeclkwd\mbox{ }v'_1\mbox{ }\sigma_1}\\
-   \ldots\\
-   \paren{\vardeclkwd\mbox{ }v_n\mbox{ }\sigma_n}\\
-   \paren{\vardeclkwd\mbox{ }v'_n\mbox{ }\sigma_n}\\
-   \paren{\constraintkwd\mbox{ }\paren{\texttt{=>}\mbox{ }
-   \paren{S_{pre}\mbox{ }v_1\mbox{ }\ldots\mbox{ }v_n}\mbox{ }
-   \paren{S\mbox{ }v_1\mbox{ }\ldots\mbox{ }v_n}}}\\
-   \paren{\constraintkwd\mbox{ }\paren{\texttt{=>}\mbox{ }
-   \paren{\texttt{and}\mbox{ }\paren{S\mbox{ }v_1\mbox{ }\ldots\mbox{ }v_n}\mbox{ }
-   \paren{S_{trans}\mbox{ }v_1\mbox{ }\ldots\mbox{ }v_n\mbox{ }v'_1\mbox{ }\ldots\mbox{ }v'_n}}\mbox{ }
-   \paren{S\mbox{ }v'_1\mbox{ }\ldots\mbox{ }v'_n}}}\\
-   \paren{\constraintkwd\mbox{ }\paren{\texttt{=>}\mbox{ }
-   \paren{S\mbox{ }v_1\mbox{ }\ldots\mbox{ }v_n}\mbox{ }
-   \paren{S_{post}\mbox{ }v_1\mbox{ }\ldots\mbox{ }v_n}}}\\
-\end{array}
-\]
+\begin{lstlisting}[language=SyGuS, numbers=none, basicstyle=\ttfamily\small]
+($\vardeclkwd$ $v_1$ $\sigma_1$)
+($\vardeclkwd$ $v'_1$ $\sigma_1$)
+$\ldots$
+($\vardeclkwd$ $v_n$ $\sigma_n$)
+($\vardeclkwd$ $v'_n$ $\sigma_n$)
+($\constraintkwd$ (=> ($S_{pre}$ $v_1$ $\ldots$ $v_n$) ($S$ $v_1$ $\ldots$ $v_n$)))
+($\constraintkwd$ (=> (and ($S$ $v_1$ $\ldots$ $v_n$) ($S_{trans}$ $v_1$ $\ldots$ $v_n$ $v'_1$ $\ldots$ $v'_n$)) ($S$ $v'_1$ $\ldots$ $v'_n$)))
+($\constraintkwd$ (=> ($S$ $v_1$ $\ldots$ $v_n$) ($S_{post}$ $v_1$ $\ldots$ $v_n$)))
+\end{lstlisting}
 where $v_1, v'_1, \ldots, v_n, v'_n$ are fresh symbols.
 
+\item \sygusinline|($\constraintchckwd$ (($x_1$ $\sigma_1$) $\ldots$ ($x_m$ $\sigma_m$)) $t_1$ $\ldots$ $t_n$ $t_{head}$)|
+
+This command adds a constrained Horn clause (CHC) constraint to the current state.
+Concretely, this command adds $m$ fresh variables $v_1, \ldots, v_m$ to the list of universal variables,
+and a constraint $((\lambda x_1,\ldots,x_m \ldotp t_{body})\ v_1\ \ldots\ v_m) \implies ((\lambda x_1,\ldots,x_m \ldotp t_{head})\ v_1\ \ldots\ v_m)$, where
+$$
+t_{body} \triangleq
+\begin{cases}
+  (t_1 \land \ldots \land t_n)  & n > 0\\
+  \truekwd                      & n = 0
+\end{cases}
+$$
+This is a generalization of the $\constraintinvkwd$ command,
+that allows for specifying problems with multiple, possibly interdependent,
+invariants to be synthesized.
+
+Let $\mathcal{S}$ be the set of symbols to be synthesized.
```

Comment:
I'm not sure this paragraph belongs here.  I think this be more appropriate where the CHC_ logic is described.

Reply:
Oh yes, now that I think about it, it makes more sense to have it there. I'll make this change.